### PR TITLE
Improve Python and additional exercise engine support

### PR DESCRIPTION
**Pull request by gadenbuie**
Improves Python exercise support in learnr. This PR replaces previous transformations at key steps in the exercise rendering process with generic functions that dispatch on the class of `exercise`, where we now add the exercise engine as a subclass of `tutorial_exercise`. This allows us to provide a generic workflow that can be extended as needed by additional engine-specific methods.

All of the generics now use the prefix `render_exercise_` since they are part of the `render_exercise()` flow.:

* `render_exercise_prepare()` replace `prepare_exercise()` and possibly modifies the `exercise` in preparation for rendering.

* `render_exercise_rmd_prep()` and `render_exercise_rmd_user()` provide a vector of source code lines that are used in the `.Rmd` rendered by `render_exercise()` to evaluate setup code (prep) and user code.

* `render_exercise_post_stage_hook()` defines hooks that are called after the prep and user `stage`s and can potentially modify `envir_prep` and `envir_result` respectively.

* `render_exercise_result()` can take all of the various exercise result pieces and may modify them all as needed to provide the final exercise result object.

One big change from the original python support implementation is that the global python environment is duplicated into `.__py__` in both `envir_prep` and `envir_result`. The new name ensures there is zero chance that we accidentally overwrite something in those environments.

Broader impact: because we're adding additional classes to the tutorial exercise object, I've incremented the exercise version number to `4` and have added a migration path in `upgrade_exercise()`.

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -53,6 +53,53 @@ duplicate_env <- function(envir, parent = parent.env(envir)) {
...
+  # extract all objects within this module
+  new_objs <- reticulate::py_get_attr(module, "__dict__")
+  # then create copy of the dictionary with all objects
+  copy <- reticulate::import("copy", convert = FALSE)
+  copy$copy(new_objs)
+}
+
+# This clears the Python environment `py`.
+#
+# It will keep important starting objects so `py` such as `r`and builtins.
```
Comment (nischalshrestha) [inline R/utils.R:84]:
```suggestion
# It will keep important initial objects such as `py` (main module), `r`, and the `builtins` module.
```

**Review by nischalshrestha — COMMENTED** [review]
This is super super exciting, and I love the generics pattern! <3 

As far as the Python bits go, this is pretty good. I like the fact that we now have a post "user" stage to include the `py` environment as `.__py__`, the `$py` always felt unclean. I just had a small suggestion which is really just fixing my bad documentation for ``

**Comment by nischalshrestha** [comment]
I _think_ I may have found a bug? So the tests currently for the new exercise structure will pass when using `mock_exercise()` because you set the engine as part of the class. 

```r
  if (version == 4) {
    class <- c(engine, class)
  }
```

However, for the latest exercise version we simply return `exercise` object without adding the new classes.

```r
  if (identical(exercise$version, current_exercise_version)) {
    return(exercise)
  }
```

That would prevent us from invoking the generic methods throughout the exercise prep, render, post stage, check process based on debugging `learnr`. And, I don't see the new classes being set for the `exercise` object.

**Comment by gadenbuie** [comment]
@nischalshrestha The classes are added in the knitr hooks, which means that the exercise objects in the cache also have those classes... but classes (like other attributes) are dropped by `append()` when we combined the user input with the full exercise object. I've fixed things in a few places to make sure that, basically no matter what, these classes are added correctly — and that the class related to the exercise engine is *always* the first class.

**Comment by nischalshrestha** [comment]
> @nischalshrestha The classes are added in the knitr hooks, which means that the exercise objects in the cache also have those classes... but classes (like other attributes) are dropped by `append()` when we combined the user input with the full exercise object. I've fixed things in a few places to make sure that, basically no matter what, these classes are added correctly — and that the class related to the exercise engine is _always_ the first class.

That's a tricky one! Good catch in where the class got lost. The `append()` has bit me similarly before. 😅

File: R/exercise.R [hunk 2]
```diff
@@ -1311,6 +1328,154 @@ local_restore_options_and_envvars <- function(.local_envir = parent.frame()) {
...
+render_exercise_rmd_user.python <- function(exercise, ...) {
+  rmd_src_user <- NextMethod()
+
+  c(
+    rmd_src_user,
+    "",
+    # this is how we get the `last_value` from the python session
+    '```{r include=FALSE}',
+    'reticulate::py_run_string("import builtins")',
+    'reticulate::py_eval("builtins._")',
```
Comment (nischalshrestha) [inline R/exercise.R:1382]: We will want to make sure to preserve the last value as a Python object:

```suggestion
    'reticulate::py_eval("builtins._", convert=FALSE)',
```

File: R/utils.R [hunk 3]
```diff
@@ -53,6 +53,54 @@ duplicate_env <- function(envir, parent = parent.env(envir)) {
...
+
+# This clears the Python environment `py`.
+#
+# It will keep important initial objects such as `py` (main module), 
+# `r` (reticulate interface to R), and the `builtins` module.
+#
+# @examples
+# reticulate::py_run_string("x = 3")
+# # this removes the `x`
+# clear_py_env(py)
```
Comment (nischalshrestha) [inline R/utils.R:90]:
```suggestion
# clear_py_env()
```

**Pull request merged**